### ISSUE TITLE
Before/after run out of order with concurrent execution

### ISSUE THREAD

**Issue by tamlyn** [DevExpress/testcafe#1909]
### Are you requesting a feature or reporting a bug?

I think it's a bug though you may tell me it's a feature.

### What is the current behavior?

When running with `concurrency` > 1, the `before` hook of one test suite can run before the `after` hook of the previous one. For example:

```
$ npx testcafe --concurrency 2 chrome test/*_test.js
 Running tests in:
 - Chrome 61.0.3163 / Mac OS X 10.11.6

 One
--> Before one
--> One: A
--> One: B
 ✓ One: a
--> One: C
 ✓ One: b
--> Before two
--> Two: A
--> After one
 ✓ One: c

 Two
--> Two: B
 ✓ Two: a
--> Two: C
 ✓ Two: b
--> After two
 ✓ Two: c
```

I understand why it happens: the second browser starts running `Two: A` before `One: C` has completed. But this is a problem when the before/after hooks are used to set up/tear down the database.

### What is the expected behavior?

I would expect it to run in the same order as when run in a single browser:

```
$ npx testcafe chrome test/*_test.js
 Running tests in:
 - Chrome 61.0.3163 / Mac OS X 10.11.6

 One
--> Before one
--> One: A
 ✓ One: a
--> One: B
 ✓ One: b
--> One: C
--> After one
 ✓ One: c

 Two
--> Before two
--> Two: A
 ✓ Two: a
--> Two: B
 ✓ Two: b
--> Two: C
--> After two
 ✓ Two: c
```

### How would you reproduce the current behavior (if this is a bug)?

Test files used to produce the above output:

```js
// test/one_test.js
fixture('One')
  .page('http://stackoverflow.com')
  .before(() => console.log('--> Before one'))
  .after(() => console.log('--> After one'))

test('One: a', () => console.log('--> One: A'))
test('One: b', () => console.log('--> One: B'))
test('One: c', () => console.log('--> One: C'))
```

```js
// test/two_test.js
fixture('Two')
  .page('http://github.com')
  .before(() => console.log('--> Before two'))
  .after(() => console.log('--> After two'))

test('Two: a', () => console.log('--> Two: A'))
test('Two: b', () => console.log('--> Two: B'))
test('Two: c', () => console.log('--> Two: C'))
```

### Specify your

```
$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.11.6
BuildVersion:	15G1421
$ npx testcafe --version
0.18.1
$ node -v
v8.6.0
```

**Comment by tamlyn** [comment]
Actually I've come up with a workaround for this that makes all browsers wait until the currently running test suite completes before starting the next one:

```js
// locker.js
let isLocked = false
const queue = []

// run before DB setup code
export async function acquire () {
  if (isLocked) {
    // another test suite is already running so wait until it finishes
    await new Promise(resolve => queue.push(resolve))
  }
  isLocked = true
}

// run at end of tear down
export function release () {
  // are there other tests waiting to run?
  if (queue.length) {
    // resolve a promise (FIFO) and start the next suite
    queue.shift()()
  }
  isLocked = false
}
```

```js
// test.js
import {acquire, release} from './locker'
fixture('My thing')
  .before(async () => {
    await acquire()
    // ... other setup
  })
  .after(() => {
    // ... other tear down
    release()
  })
```

I'm still interested in the dev teams thoughts about the default behaviour though, or any more elegant solutions.

**Comment by AndreyBelym** [comment]
Hello @tamlyn!

I got your point, but the expected log which you wrote looks like just a log with `concurrency === 1`. I think, with `concurrency > 1`, it will be look more like:
```
 One
--> Before one
--> One: A
--> One: B
 ✓ One: a
--> One: C
 ✓ One: b
--> After one
 ✓ One: c

 Two
--> Before two
--> Two: A
--> Two: B
 ✓ Two: a
--> Two: C
 ✓ Two: b
--> After two
 ✓ Two: c
``` 

So some tests are started before previous was completed.

We discussed it and it seems like a good idea. We can create a barrier, that will ensure, that `before` hook and tests from the second fixture will not run before the `after` hook from the first fixture will be finished. But it will not stop starting tests from the second fixture if fixtures don't have hooks, because it will significantly increase overall tests execution time.
